scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/CascadingBackend.scala

```diff
@@ -303,61 +303,61 @@ object CascadingBackend {
     RichPipe.setPipeDescriptions(p, ordered ::: unordered)
   }
 
-  final def toPipe[U](p: TypedPipe[U], fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
-
-    /**
-     * Here are configurable optimization rules
-     */
-    val getHashJoinAutoForceRight =
-      mode match {
-        case h: HadoopMode =>
-          val config = Config.fromHadoop(h.jobConf)
-          config.getHashJoinAutoForceRight
-        case _ => false
-      }
-    val forceHash = if (getHashJoinAutoForceRight) OptimizationRules.ForceToDiskBeforeHashJoin else Rule.empty[TypedPipe]
-    /**
-     * These are rules we should apply to any TypedPipe before handing
-     * to cascading. These should be a bit conservative in that they
-     * should be highly likely to improve the graph.
-     */
-    val phases = List(
-      // phase 0, add explicit forks
-      OptimizationRules.AddExplicitForks,
-      // phase 1, compose flatMap/map, move descriptions down etc...
-      Rule.orElse(List(
-        OptimizationRules.ComposeMap,
-        OptimizationRules.ComposeFilter,
-        OptimizationRules.ComposeWithOnComplete,
-        OptimizationRules.DescribeLater)),
-      // phase 2, combine different kinds of map operations into flatMaps
-      Rule.orElse(List(
-        OptimizationRules.ComposeMapFlatMap,
-        OptimizationRules.ComposeFilterFlatMap,
-        OptimizationRules.ComposeFilterMap,
-        OptimizationRules.ComposeFlatMap,
-        OptimizationRules.EmptyIsOftenNoOp,
-        OptimizationRules.EmptyIterableIsEmpty)),
-      // phase 3, add any explicit forces to the optimized graph
-      Rule.orElse(List(
-        forceHash,
-        OptimizationRules.RemoveDuplicateForceFork))
-      )
-
+  /**
+   * These are rules we should apply to any TypedPipe before handing
+   * to cascading. These should be a bit conservative in that they
+   * should be highly likely to improve the graph.
+   */
+  def defaultOptimizationRules(config: Config): Seq[Rule[TypedPipe]] = {
     /**
      * TODO:
      * we need to have parity for the normal optimizations
      * scalding has been applying in 0.17
      *
      */
+    def std(forceHash: Rule[TypedPipe]) = OptimizationRules.standardMapReduceRules :+
+      // add any explicit forces to the optimized graph
+      Rule.orElse(List(
+        forceHash,
+        OptimizationRules.RemoveDuplicateForceFork)
+      )
+
+    config.getOptimizationPhases match {
+      case Some(tryPhases) => tryPhases.get.phases
+      case None =>
+        val force =
+          if (config.getHashJoinAutoForceRight) OptimizationRules.ForceToDiskBeforeHashJoin
+          else Rule.empty[TypedPipe]
+        std(force)
+    }
+  }
+
+  final def toPipe[U](p: TypedPipe[U], fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
 
+    val phases = defaultOptimizationRules(
```

Comment:
If we are optimizing on toPipe does the optimization rules in the Execution have any effect?

Reply:
we are careful to avoid calling `toPipe` in Execution. We call `toPipeUnoptimized`. This is because we want to do the optimization on the whole graph (multiple sinks) that Execution can see. This is likely to be better and allow more aggressive optimization than the per-pipe approach of `.write`.

Reply:
Do we not hit toPipe going in and out of groupBy's?